### PR TITLE
ADBM-2434: Fix wal filtering when there is no unfinished record at the end of the file

### PR DESCRIPTION
**Pull request by KnightMurloc**
Fix wal filtering when there is no unfinished record at the end of the file

When there is no unfinished record at the end of the WAL segment, the next
record will be read, but when the next page is requested, it will be found that
the segment file has ended. However, since the current step has already been set
to stepBeginOfRecord, it was mistakenly assumed that there is an unfinished
record at the end that needs to be written. And since this is also the beginning
of the page, an attempt will be made to write down its header, which is missing,
which led to a crash.

Fix this by improving the condition for detecting the presence of an unfinished
record at the end of the file. Now it is assumed that the current record is
incomplete only if we have at least started reading the header. A test has also
been added for another edge case if the WAL switch record is located at the very
end of the page.

### CODE REVIEW

File: src/common/walFilter/walFilter.c [hunk 1]
```diff
@@ -513,8 +513,8 @@ walFilterProcess(THIS_VOID, const Buffer *const input, Buffer *const output)
 
     if (input == NULL)
     {
-        // We have an incomplete record at the end
-        if (this->currentStep != noStep)
+        // We have an incomplete record at the end and we already read something
```
Comment (red1452) [inline src/common/walFilter/walFilter.c]: "we have already read"

**Review by red1452 — DISMISSED** [review]

**Review by red1452 — APPROVED** [review]

**Review by dkovalev1 — APPROVED** [review]
